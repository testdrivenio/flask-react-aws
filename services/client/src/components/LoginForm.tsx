import React from "react";
import { Formik, Form, Field } from "formik";
import { z } from "zod";
import {
  Box,
  Button,
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  VStack,
  Heading,
} from "@chakra-ui/react";
import { Navigate } from "react-router-dom";

const validationSchema = z.object({
  email: z.string().email("Enter a valid email"),
  password: z.string().min(1, "Password is required"),
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type FormValues = z.infer<typeof validationSchema>;

interface LoginFormProps {
  onSubmit: (values: { email: string; password: string }) => Promise<void>;
  isAuthenticated: () => boolean;
}

const LoginForm: React.FC<LoginFormProps> = ({ onSubmit, isAuthenticated }) => {
  if (isAuthenticated()) {
    return <Navigate to="/" replace />;
  }
  return (
    <Box maxWidth="400px" margin="100px auto 0">
      <Heading as="h1" size="xl" textAlign="center" mb={6}>
        Log In
      </Heading>
      <Formik
        initialValues={{
          email: "",
          password: "",
        }}
        onSubmit={(values, { setSubmitting, resetForm }) => {
          onSubmit(values);
          resetForm();
          setSubmitting(false);
        }}
        validate={(values) => {
          try {
            validationSchema.parse(values);
            return {};
          } catch (error) {
            return (error as z.ZodError).formErrors.fieldErrors;
          }
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form>
            <VStack spacing={4}>
              <Field name="email">
                {({ field }) => (
                  <FormControl isInvalid={!!(errors.email && touched.email)}>
                    <FormLabel htmlFor="email">Email</FormLabel>
                    <Input
                      {...field}
                      id="email"
                      type="email"
                      placeholder="Enter an email address"
                    />
                    <FormErrorMessage>{errors.email}</FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="password">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.password && touched.password)}
                  >
                    <FormLabel htmlFor="password">Password</FormLabel>
                    <Input
                      {...field}
                      id="password"
                      type="password"
                      placeholder="Enter a password"
                    />
                    <FormErrorMessage>{errors.password}</FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Button
                type="submit"
                colorScheme="blue"
                width="full"
                isLoading={isSubmitting}
                bg="green.400"
                _hover={{ bg: "green.500" }}
              >
                Log In
              </Button>
            </VStack>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default LoginForm;
