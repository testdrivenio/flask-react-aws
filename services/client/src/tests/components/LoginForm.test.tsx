import React from "react";
import { render, screen } from "../test-utils";
import { describe, it, expect, vi } from "vitest";
import { expect as expectVitest } from "vitest";
import { BrowserRouter as Router } from "react-router-dom";
import LoginForm from "../../components/LoginForm";

const mockProps = {
  onSubmit: vi.fn(),
  isAuthenticated: vi.fn().mockReturnValue(false),
};

const renderWithRouter = (ui: React.ReactElement, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);
  return render(ui, { wrapper: Router });
};

describe("LoginForm", () => {
  it("renders properly", () => {
    renderWithRouter(<LoginForm {...mockProps} />);
    const heading = screen.getByRole("heading", { name: "Log In" });
    expect(heading.tagName.toLowerCase()).toBe("h1");
  });

  it("renders with default props", () => {
    renderWithRouter(<LoginForm {...mockProps} />);

    const emailInput = screen.getByLabelText("Email") as HTMLInputElement;
    expect(emailInput.type).toBe("email");
    expect(emailInput.value).toBe("");

    const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;
    expect(passwordInput.type).toBe("password");
    expect(passwordInput.value).toBe("");

    const submitButtons = screen.getAllByRole("button", { name: "Log In" });
    expect(submitButtons[0].textContent).toBe("Log In");
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<LoginForm {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders login form when not authenticated", () => {
    const { container } = renderWithRouter(<LoginForm {...mockProps} />);
    const heading = container.querySelector("h1");
    expectVitest(heading?.textContent).toBe("Log In");
  });

  it("redirects when authenticated", () => {
    const authenticatedProps = {
      ...mockProps,
      isAuthenticated: vi.fn().mockReturnValue(true),
    };
    renderWithRouter(<LoginForm {...authenticatedProps} />);
    expect(window.location.pathname).toBe("/");
  });
});
