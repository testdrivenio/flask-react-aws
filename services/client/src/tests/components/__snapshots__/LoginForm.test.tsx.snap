// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > renders 1`] = `
<DocumentFragment>
  <div
    class="css-17k3x8d"
  >
    <h1
      class="chakra-heading css-ov0hf1"
    >
      Log In
    </h1>
    <form
      action="#"
    >
      <div
        class="chakra-stack css-1xb2ccj"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="email"
            id="field-:r4:-label"
          >
            Email
          </label>
          <input
            class="chakra-input css-0"
            id="email"
            name="email"
            placeholder="Enter an email address"
            type="email"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="password"
            id="field-:r5:-label"
          >
            Password
          </label>
          <input
            class="chakra-input css-0"
            id="password"
            name="password"
            placeholder="Enter a password"
            type="password"
            value=""
          />
        </div>
        <button
          class="chakra-button css-1lmcr0o"
          type="submit"
        >
          Log In
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
